packages/replay/src/auth.ts

```diff
@@ -263,3 +264,128 @@ export async function maybeAuthenticateUser(options: Options = {}) {
+  const resp = await query(
+    "AuthInfo",
+    `
+        query AuthInfo {
+          viewer {
+            user {
+              id
+            }
+          }
+          auth {
+            workspaces {
+              edges {
+                node {
+                  id
+                }
+              }
+            }
+          }
+        }
+      `,
+    undefined,
+    key
+  );
+
+  if (resp.errors) {
+    throw {
+      id: "graphql-error",
+      message: resp.errors
+        .map((e: any) => e.message)
+        .filter(Boolean)
+        .join(", "),
+    };
+  }
+
+  const response = resp.data as {
+    viewer: {
+      user: {
+        id: string | null;
+      } | null;
+    };
+    auth: {
+      workspaces: {
+        edges: {
+          node: {
+            id: string;
+          };
+        }[];
+      };
+    };
+  };
+
+  const encodedUserId = response?.viewer?.user?.id;
+  const encodedWorkspaceId = response?.auth?.workspaces?.edges?.[0]?.node?.id;
+
+  return {
+    userId: encodedUserId ? parseId(encodedUserId) : null,
+    workspaceId: encodedWorkspaceId ? parseId(encodedWorkspaceId) : null,
+  };
+}
+
+function getAuthInfoCachePath(options: Options = {}) {
+  const directory = getDirectory(options);
+  return path.resolve(path.join(directory, "profile", "authInfo.json"));
+}
+
+function authInfoCacheKey(key: string) {
+  return createHash("sha256").update(key).digest("hex");
+}
+
+async function writeAuthInfoCache(
+  key: string,
+  authInfo: { userId: string | null; workspaceId: string | null },
+  options: Options = {}
+) {
+  const cachePath = getAuthInfoCachePath(options);
+  await mkdir(path.dirname(getAuthInfoCachePath(options)), { recursive: true });
+  const cache = {
+    [authInfoCacheKey(key)]: authInfo,
+  };
+  await writeFile(cachePath, JSON.stringify(cache, undefined, 2), { encoding: "utf-8" });
+}
+
+async function readAuthInfoCache(key: string, options: Options = {}) {
+  try {
+    const cachePath = getAuthInfoCachePath(options);
+    const cacheJson = await readFile(cachePath, { encoding: "utf-8" });
+    const cache = JSON.parse(cacheJson);
+    return cache[authInfoCacheKey(key)];
+  } catch (e) {
+    debug("Failed to read auth info cache: %o", e);
+    return;
+  }
+}
+
+export async function initLDContextFromKey(options: Options = {}) {
+  const apiKey = options.apiKey ?? (await readToken(options));
+  if (!apiKey) {
+    return;
+  }
+
+  let authInfo = await readAuthInfoCache(apiKey, options);
+  if (!authInfo) {
+    debug("Fetching auth info from server");
+    authInfo = await getAuthInfo(apiKey);
+    if (authInfo) {
+      await writeAuthInfoCache(apiKey, authInfo, options);
+    }
+  }
+
+  if (!authInfo) {
+    return;
+  }
+
+  if (authInfo.userId) {
+    await launchDarkly.identify({ type: "user", id: authInfo.userId });
+  } else if (authInfo.workspaceId) {
+    await launchDarkly.identify({ type: "workspace", id: authInfo.workspaceId });
+  }
+}
```

Comment:
This might be better placed in the `./launchdarkly` module and export a fn from here that is responsible for reading (and possibly updating) the cache.

Reply:
This function depends both on auth graphql calls and then calls launchdarkly. I discussed it with @Andarist felt it's more dependent on auth and decided to keep it here.